src/plugins/intel_cpu/src/emitters/snippets/x64/kernel_executors/brgemm_base.hpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (C) 2020-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "openvino/core/type/element_type.hpp"
+
+#include "cpu/x64/cpu_isa_traits.hpp"
+
+#include "emitters/plugin/x64/jit_emitter.hpp"
+#include "emitters/snippets/jit_snippets_call_args.hpp"
+#include "emitters/snippets/cpu_kernel_executor_table.hpp"
+#include <cpu/x64/brgemm/brgemm.hpp>
+
+#include "snippets/lowered/loop_manager.hpp"
+#include "snippets/lowered/loop_info.hpp"
+
+namespace ov {
+namespace intel_cpu {
+
+struct BrgemmBaseKernelConfig {
+public:
+    virtual ~BrgemmBaseKernelConfig() = default;
+
+    bool is_completed() const;
+    bool is_empty() const;
+    void update(dnnl_dim_t M, dnnl_dim_t N, dnnl_dim_t K, dnnl_dim_t LDA, dnnl_dim_t LDB, dnnl_dim_t LDC, float beta);
+
+    bool operator==(const BrgemmBaseKernelConfig& rhs) const;
+    bool operator!=(const BrgemmBaseKernelConfig& rhs) const {return !(*this == rhs);}
+
+    size_t hash() const { return m_hash; }
+
+    dnnl_data_type_t get_dt_in0() const { return m_static_params->dt_in0; }
+    dnnl_data_type_t get_dt_in1() const { return m_static_params->dt_in1; }
+
+    dnnl::impl::cpu::x64::cpu_isa_t get_isa() const { return m_static_params->isa; }
+    float get_beta() const { return m_beta; }
+
+    dnnl_dim_t get_M() const { return m_M; }
+    dnnl_dim_t get_N() const { return m_N; }
+    dnnl_dim_t get_K() const { return m_K; }
+
+    dnnl_dim_t get_LDA() const { return m_LDA; }
+    dnnl_dim_t get_LDB() const { return m_LDB; }
+    dnnl_dim_t get_LDC() const { return m_LDC; }
+
+#ifdef SNIPPETS_DEBUG_CAPS
+    std::string to_string() const;
+#endif
+
+protected:
+    struct StaticBaseParams {
+        StaticBaseParams(const element::Type& in0_dtype, const element::Type& in1_dtype, dnnl::impl::cpu::x64::cpu_isa_t primitive_isa);
+        virtual ~StaticBaseParams() = default;
+
+        const dnnl_data_type_t dt_in0 {dnnl_f32}, dt_in1 {dnnl_f32};
+        const dnnl::impl::cpu::x64::cpu_isa_t isa {dnnl::impl::cpu::x64::isa_undef};
+
+        virtual size_t hash() const = 0;
+
+        bool operator==(const StaticBaseParams& rhs) const;
+        bool operator!=(const StaticBaseParams& rhs) const { return !(*this == rhs); }
+#ifdef SNIPPETS_DEBUG_CAPS
+        std::string to_string() const;
+#endif
+    protected:
+        virtual size_t compute_hash() const;
```

Comment:
The idea was to save consistency of the field `m_hash` in the classes. If I set `const size_t m_hash` in base class - derived classes cannot update this value in the own ctor. However, they have the different attributes which should affect to hash value calculated using base common attributes.
I cannot init  `const size_t m_hash` in base class constructor using virtual method `compute_hash()` as `m_hash(compute_hash())` - because derived classes haven't been inited and there will be called `StaticBaseParams::compute_hash()` instead of derived classes methods.

For sure, we may store `m_hash` in base class `StaticBaseParams` - but it will be non constant field to have posibility to init it in derived classes.

Reply:
Ok, I see what you are trying to achieve, but from my perspective an additional virtual method in the base class is a high price for a constant field 🙂 
It feels like we are trying to sit on two chairs here: have both virtual compute_hash and const m_hash.
When we create virtual compute_hash, we declare smth like: "The base class will need to compute hash, but doesn't know how". The paradox here is that the base class never actually calls `compute_hash` and it's not a part of its public interface. The virtual compute_hash would make sense if we call it from `hash()` for example: if m_hash is not initialized => compute hash, if it is => return it. But it's not the case because `hash()` is also purely virtual.

What do you think about moving `m_hash` to the base class + making it const?
1) We add a `seed` argument to the base class constructor, it will then be used to initialize m_hash via `hash_combine`
2) Every class creates a static compute_hash(agr1, arg2, ...) function.
3) Then Base & derived class constructors would look smth like:
```
StaticBaseParams::StaticBaseParams(const element::Type& in0_dtype, const element::Type& in1_dtype,
                                                           dnnl::impl::cpu::x64::cpu_isa_t primitive_isa, **size_t seed = 0**)
    : dt_in0(DTYPE_CAST(in0_dtype)), dt_in1(DTYPE_CAST(in1_dtype)), isa(primitive_isa), **m_hash(compute_hash(seed, dt_in0, dt_in1, isa)**) {}
StaticParams::StaticParams(const element::Type& in0_dtype, const element::Type& in1_dtype,
                                                  dnnl::impl::cpu::x64::cpu_isa_t primitive_isa)
    : StaticBaseParams(in0_dtype, in1_dtype, primitive_isa, **compute_hash(inner_k_blk, vnni_factor)**)
```
This way we'll have a clearer interface: no virtual functions with similar names and non-obvious differences (in fact, even `hash()` will be static), `m_hash` will be constant.

Reply:
Great idea! Thank you for the detailed thoughts! Implemented 👍🏼